appengine/flexible/TwelveFactor/Services/CloudStorageFileProvider.cs

```diff
@@ -0,0 +1,259 @@
+                        }
+                    } 
+                    catch (Exception e) 
+                    {
+                        _logger.LogError(0, e, "Exception while watching {0}",
+                            filter);
+                    }
+                }
+            });
+            return token;            
+        }            
+    }
+
+    class CloudStorageChangeToken : IChangeToken
+    {
+        List<ChangeCallback> _callbacks = new List<ChangeCallback>();
+        bool _hasChanged = false;
+
+        public bool HasChanged { 
+            get { return _hasChanged; }
+            set {
+                _hasChanged = true;
+                foreach (var callback in _callbacks) {
+                    callback.Invoke();
+                }
+            }
+        }
+
+        public bool ActiveChangeCallbacks => true;
+
+        public IDisposable RegisterChangeCallback(Action<object> callback,
+            object state)
+        {
+            var disposable = new ChangeCallback() {
+                Callback = callback,
+                State = state
+            };
+            _callbacks.Add(disposable);
+            return disposable;
+        }
+
+        class ChangeCallback : IDisposable {
+            public Action<object> Callback { get; set; }
+            public object State {get; set;}
+
+            private object _lock = new object();
+            private bool _done = false;
+            public void Dispose()
+            {
+                lock(_lock) {
+                    if (!_done) {
+                        _done = true;
+                    }
+                }
+            }
+
+            public void Invoke() {
+                lock(_lock) {
+                    if (!_done) {
+                        _done = true;
+                        Callback(State);
+                    }
+                }                
+            }
+        };
+    }
+
+    class CloudStorageFileInfo : IFileInfo
+    {
+        readonly Google.Apis.Storage.v1.Data.Object _cloudStorageObject;
+        readonly StorageClient _storage;
+        readonly ILogger _logger;
+
+        public CloudStorageFileInfo(
+            Google.Apis.Storage.v1.Data.Object cloudStorageObject,
+            StorageClient storageClient, ILogger logger)
+        {
+            _cloudStorageObject = cloudStorageObject;
+            _storage = storageClient;
+            _logger = logger;
+        }
+
+        public bool Exists => true;
+
+        public long Length => (long) _cloudStorageObject.Size.Value;
+
+        public string PhysicalPath => null;
+
+        public string Name => _cloudStorageObject.Name;
+
+        public DateTimeOffset LastModified => 
+            _cloudStorageObject.Updated.Value;
+
+        public bool IsDirectory => false;
+
+        public Stream CreateReadStream()
+        {
+            // Json config files are small, and will easily fit in memory.
+            var stream = new MemoryStream();

```

Comment:
Nit, you could cache the contents in the `CloudStorageFileInfo` instance, so you avoid downloading the contents every time the `CreateReadStream()` method is called. Of course this might not make any difference if the method is only called once.

Reply:
Is not really about the size, is more about the latency.